File: ios/Runner/ReelsPlayerPlugin.swift
import Flutter
import AVFoundation

// ═══════════════════════════════════════════════════════════════════════════
// ReelsPlayerPlugin
//
// Bridges Flutter MethodChannel "reels_player" <-> ReelsPlayerPool.
// Mirrors Android's ReelsPlayerPlugin exactly — same method names,
// same argument keys, same error codes.
//
// Registration: called from AppDelegate BEFORE GeneratedPluginRegistrant.
// ═══════════════════════════════════════════════════════════════════════════

class ReelsPlayerPlugin: NSObject {

    static let channelName = "reels_player"

    private let pool: ReelsPlayerPool
    private var channel: FlutterMethodChannel!

    init(pool: ReelsPlayerPool, messenger: FlutterBinaryMessenger) {
        self.pool = pool
        super.init()

        channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
        channel.setMethodCallHandler(handle)

        // Give the pool a reference so it can call firstFrameReady back to Dart
        pool.methodChannel = channel
    }

    // ─────────────────────────────────────────────────────────────────────────
    // MARK: - Method call handler
    // ─────────────────────────────────────────────────────────────────────────

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // All calls are already on the main thread (Flutter platform channel guarantee)
        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {

        case "prepare":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            guard let url = args["url"] as? String else {
                return result(FlutterError(code: "ARG", message: "url missing", details: nil))
            }
            let token = args["token"] as? String ?? ""
            pool.prepare(slot: slot, url: url, token: token)
            result(nil)

        case "switchSurface":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            pool.switchSurface(toSlot: slot)
            result(nil)

        case "play":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            pool.play(slot: slot)
            result(nil)

        case "pause":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            pool.pause(slot: slot)
            result(nil)

        case "setVolume":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            let volume = (args["volume"] as? Double).map { Float($0) } ?? 1.0
            pool.setVolume(slot: slot, volume: volume)
            result(nil)

        case "seekTo":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            let posMs = args["positionMs"] as? Int ?? 0
            pool.seekTo(slot: slot, positionMs: posMs)
            result(nil)

        case "release":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            pool.release(slot: slot)
            result(nil)

        case "onFirstFrame":
            guard let slot = args["slot"] as? Int else {
                return result(FlutterError(code: "ARG", message: "slot missing", details: nil))
            }
            pool.setOnFirstFrameListener(slot: slot)
            result(nil)

        case "releaseAll":
            pool.releaseAll()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/app/src/main/kotlin/com/example/innovator/reels/ReelsPlayerPlugin.kt
package com.innovation.innovator.reels

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StandardMessageCodec

/**
 * Bridges Flutter MethodChannel "reels_player" <-> ReelsPlayerPool.
 * No fallbackUrl handling needed here — setEnableDecoderFallback(true)
 * in the pool handles all codec issues transparently.
 */
class ReelsPlayerPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var pool: ReelsPlayerPool
    private val mainHandler = android.os.Handler(android.os.Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        pool    = ReelsPlayerPool(binding.applicationContext)
        channel = MethodChannel(binding.binaryMessenger, "reels_player")
        channel.setMethodCallHandler(this)
        binding.platformViewRegistry.registerViewFactory(
            "reels_surface_view",
            ReelsSurfaceViewFactory(pool, StandardMessageCodec.INSTANCE)
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        pool.releaseAll()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            "prepare" -> {
                val slot  = call.argument<Int>("slot")    ?: return result.error("ARG","slot missing",null)
                val url   = call.argument<String>("url")  ?: return result.error("ARG","url missing",null)
                val token = call.argument<String>("token") ?: ""
                pool.prepare(slot, url, token)
                result.success(null)
            }

            "switchSurface" -> {
                val slot = call.argument<Int>("slot") ?: return result.error("ARG","slot missing",null)
                pool.switchSurface(slot)
                result.success(null)
            }

            "play" -> {
                val slot = call.argument<Int>("slot") ?: return result.error("ARG","slot missing",null)
                pool.play(slot)
                result.success(null)
            }

            "pause" -> {
                val slot = call.argument<Int>("slot") ?: return result.error("ARG","slot missing",null)
                pool.pause(slot)
                result.success(null)
            }

            "setVolume" -> {
                val slot   = call.argument<Int>("slot")          ?: return result.error("ARG","slot missing",null)
                val volume = call.argument<Double>("volume")?.toFloat() ?: 1f
                pool.setVolume(slot, volume)
                result.success(null)
            }

            "seekTo" -> {
                val slot  = call.argument<Int>("slot")      ?: return result.error("ARG","slot missing",null)
                val posMs = call.argument<Int>("positionMs")?.toLong() ?: 0L
                pool.seekTo(slot, posMs)
                result.success(null)
            }

            "release" -> {
                val slot = call.argument<Int>("slot") ?: return result.error("ARG","slot missing",null)
                pool.release(slot)
                result.success(null)
            }

            "onFirstFrame" -> {
                val slot = call.argument<Int>("slot") ?: return result.error("ARG","slot missing",null)
                pool.setOnFirstFrameListener(slot) {
                    mainHandler.post { channel.invokeMethod("firstFrameReady", mapOf("slot" to slot)) }
                }
                result.success(null)
            }

            "releaseAll" -> { pool.releaseAll(); result.success(null) }

            else -> result.notImplemented()
        }
    }
}